server/routes/materials.ts

```diff
@@ -33,75 +54,168 @@ router.get('/:id/users', async (req: Request, res: Response) => {
     return res.status(404).json({ error: 'Material not found' })
 })
 
-router.post('/', async (req: Request, res: Response) => {
-    const { title, description, price, file, type }: MaterialInputs = req.body
-
-    if (!title || !description || !price || !file || !type) {
+router.post('/', upload.single('file'), async (req: Request, res: Response) => {
+    try {
+      const { title, description, price, type }: MaterialInputs = req.body;
+  
+      if (!title || !description || !price || !req.file || !type) {
         return res.status(400).json({
-            error: 'Missing required fields (title, description, price, file, type)',
-        })
+          error: 'Missing required fields (title, description, price, file, type)',
+        });
+      }
+  
+      // TODO: Change to logged user
+      const author = new mongoose.Types.ObjectId('60d5ecb44b930ac130e82d7e')
+  
+      const blob = bucket.file(`${uuidv4()}-${req.file.originalname}`)
+      const blobStream = blob.createWriteStream({
+        metadata: {
+          contentType: req.file.mimetype,
+        },
+      });
+  
+      blobStream.on('error', (err) => {
+        console.error('Error al subir el archivo:', err)
+        res.status(500).json({ error: 'Error uploading file.' })
+      });
+  
+      blobStream.on('finish', async () => {
+        const publicUrl = `https://storage.googleapis.com/${bucketName}/${blob.name}`
+  
+        const newMaterial = Material.build({
+          title,
+          description,
+          price,
+          author,
+          purchasers: [],
+          file: publicUrl,
+          type,
+        });
+  
+        try {
+          const savedMaterial = await newMaterial.save()
+          res.status(201).json(savedMaterial)
+        } catch (error) {
+          console.error('Error al guardar el material en la base de datos:', error)
+          res.status(500).json({ error: 'Error saving material.' })
+        }
+      });
+  
+      blobStream.end(req.file.buffer)
+    } catch (error) {
+      console.error('Error en la solicitud:', error)
+      res.status(500).json({ error: 'Internal Server Error' })
     }
+  })
 
-    //TODO: Change to logged user
-    //const author = req.user ? res.user.id : null;
-    const author = new mongoose.Types.ObjectId('60d5ecb44b930ac130e82d7e')
-
-    const newMaterial = Material.build({
-        title,
-        description,
-        price,
-        author,
-        purchasers: [],
-        file,
-        type,
-    })
-
-    const savedMaterial = await newMaterial.save()
-
-    res.status(201).json(savedMaterial)
-})
-
-router.put('/:id', async (req: Request, res: Response) => {
+router.put('/:id',  upload.single('file'), async (req: Request, res: Response) => {
```

Comment:
The same happens as in the POST operation, you must validate that the inserted fields (title, description, price, type) are correct before uploading the file. Right now if I try to change the file with an incorrect field the material is not updated in the database but the old file is deleted and the new one is uploaded in the Google Cloud bucket.